src/lib.rs

```diff
@@ -242,40 +258,46 @@
 //! extern crate nb;
 //!
 //! use hal::prelude::*;
+//! use futures::{
+//!     future,
+//!     Async,
+//!     Future,
+//! };
+//! use futures::future::Loop;
 //! use stm32f103xx_hal_impl::{Serial, Timer};
 //!
 //! /// `futures` version of `Timer.wait`
 //! ///
 //! /// This returns a future that must be polled to completion
-//! fn wait<T>(timer: &T) -> impl Future<Item = (), Error = !>
+//! fn wait<T>(timer: T) -> impl Future<Item = (), Error = !>
```

Comment:
With this signature the function will destroy the timer object. The pattern that the futures ecosystem uses to work around this issue is to return the input by value in `Item`. Thus the return type should be `impl Future<Item = T, Error = !>`.

Reply:
I don't think this pattern will work with `poll_fn`. Here's what I got now:
``` rust
fn wait<T>(timer: T) -> impl Future<Item = T, Error = !>
where
    T: hal::Timer,
{
    future::poll_fn(|| {
        try_nb!(timer.wait());
        Ok(Async::Ready(timer))
    })
}
```
And here's how it fails:
```
error[E0507]: cannot move out of captured outer variable in an `FnMut` closure
  --> src/lib.rs:29:25
   |
23 | fn wait<T>(timer: T) -> impl Future<Item = T, Error = !>
   |            ----- captured outer variable
...
29 |         Ok(Async::Ready(timer))
   |                         ^^^^^ cannot move out of captured outer variable in an `FnMut` closure
```
Which makes sense, as `poll_fn` takes an `FnMut` closure.
Of course I could re-create `poll_fn`/`PollFn`, but this seems a bit to involved for an example like this. What do you think?

Reply:
Yeah, I *think* you'll have to use [loop_fn](https://docs.rs/futures/0.1.16/futures/future/fn.loop_fn.html) instead.

Reply:
Thanks for the help! I think I got it now.